### ISSUE TITLE
[raising-bad-type] false negative if raised before

### ISSUE THREAD

**Issue by stanislavlevin** [PyCQA/pylint#4772]
### Bug description

```python
# pylint: disable=missing-docstring,raise-missing-from,no-self-use
def handle_exception():
    print("do anything there")
    raise Exception("oops")

try:
    1 // 0
except ZeroDivisionError:
    raise handle_exception()

class MyClass:
    def handle_exception(self):
        print("do anything there")
        raise Exception("oops")

    def my_method(self):
        try:
            1//0
        except ZeroDivisionError:
            raise self.handle_exception()
```
Note: the error depends on the presence of any statement in the body of `handle_exception` before the raise (for example, no error if remove `print`)

### Command used

```shell
python3 -m pylint a.py
```

### Pylint output

```shell
************* Module a
a.py:9:4: E0702: Raising NoneType while only classes or instances are allowed (raising-bad-type)
a.py:20:12: E0702: Raising NoneType while only classes or instances are allowed (raising-bad-type)

------------------------------------------------------------------
```

### Expected behavior

There should be no error.

### Pylint version

```shell
pylint 2.9.6
astroid 2.6.5
Python 3.9.6 (default, Jun 29 2021, 10:42:27) 
[GCC 10.2.1 20210313 (ALT Sisyphus 10.2.1-alt3)]
```

### OS / Environment

ALTLinux Sisyphus

**Comment by stanislavlevin** [comment]
Another example with the same result (a.py):
```python
# pylint: disable=missing-docstring,deprecated-module
from optparse import OptionParser

def main():
    parser = OptionParser()
    raise parser.error("oops")

if __name__ == "__main__":
    main()

```

optparse.py:
```python
    def error(self, msg):
        """error(msg : string)

        Print a usage message incorporating 'msg' to stderr and exit.
        If you override this in a subclass, it should not return -- it
        should either exit or raise an exception.
        """
        self.print_usage(sys.stderr)
        self.exit(2, "%s: error: %s\n" % (self.get_prog_name(), msg))
```

**Comment by Pierre-Sassoulas** [comment]
I think you should do `self.handle_exception()` instead of `raise self.handle_exception()`, same thing for `parser.error()`.

**Comment by stanislavlevin** [comment]
IMO, It's OK case.
Please see details about motivation:
https://github.com/freeipa/freeipa/commit/f60b2c590684c8ffc2c6ce2efa7a99f0193f42a1

**Comment by Pierre-Sassoulas** [comment]
I'm nitpicking but the pylint warning is true, it's just that there is always an exception raised before reaching the problem.

> It makes the code easier to understand, the control flow more obvious and helps static analyzers.

I think it's true, but as pylint do not have real control flow "checking that a function always raise" is not trivial for us.
